compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt

```diff
@@ -89,6 +91,32 @@ class Fir2IrClassifierStorage(
                     ?: createIrTypeParameterWithoutBounds(original, index, context)
             }
         }
+        if (owner is FirCallableDeclaration<*>) {
+            val typeParameterInReturnType = owner.returnTypeRef.retrieveTypeParameter()
+            if (typeParameterInReturnType != null) {
+                getCachedIrTypeParameter(typeParameterInReturnType)
+                    ?: createIrTypeParameterWithoutBounds(typeParameterInReturnType, index = 0)
+            }
+        }
+        if (owner is FirFunction<*>)
+            owner.valueParameters.forEachIndexed { index, valueParameter ->
+                val typeParameterInValueParameter = valueParameter.returnTypeRef.retrieveTypeParameter()
+                if (typeParameterInValueParameter != null) {
+                    getCachedIrTypeParameter(typeParameterInValueParameter)
+                        ?: createIrTypeParameterWithoutBounds(typeParameterInValueParameter, index)
```

Comment:
`index` is incorrect here, it's value parameter index, but should be type parameter index.